RabbitMQ and OSLO: Start RPC server/client and restart one/all controllers

alex7c4 opened the pull request:
QA-2290 - Start RabbitMQ RPC server and client and restart RabbitMQ on one controller
QA-2291 - Start RabbitMQ RPC server and client and restart RabbitMQ on all controllers one-by-one
QA-2293 - Start RabbitMQ RPC server and client and restart non-primary controller
QA-2294 - Start RabbitMQ RPC server and client and restart primary controller

Commit:
RabbitMQ and OSLO: Start RPC server/client and restart one/all controllers

QA-2291 - Start RabbitMQ RPC server and client and restart RabbitMQ on all controllers one-by-one
QA-2290 - Start RabbitMQ RPC server and client and restart RabbitMQ on one controller

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +235,32 @@ def consume_msg(remote, cfg_file_path):
+    logger.debug('Start [oslo_msg_check_server] on %s.' % remote.host)
+    cmd = 'oslo_msg_check_server --nodebug --config-file {0}'.format(
+        cfg_file_path)
+    remote.execute_async(cmd)
```
This may leave open ssh transport channels. Why you didn't used `check_call` or  `execute`?

alex7c4 replied:
This is because I need to keep server/client running on node and continue test-code execution.
`check_call` or `execute` are waiting for command to finish it's execution.
But in my case server/client is running and keep posting information in console.

gdyuldin replied:
`<&- >/dev/null 2>&1 &` - add it to end of command and it will run in background even if you close SSH connection.

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +235,32 @@ def consume_msg(remote, cfg_file_path):
+    logger.debug('Start [oslo_msg_check_client] on %s.' % remote.host)
+    cmd = 'oslo_msg_check_client --nodebug --config-file {0}'.format(
+        cfg_file_path)
+    remote.execute_async(cmd)
```
This may leave open ssh transport channels. Why you didn't used `check_call` or  `execute`?

alex7c4 replied:
Please see my comment above

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +235,32 @@ def consume_msg(remote, cfg_file_path):
+    cmd = 'curl -w %{http_code} -s -o /dev/null http://{host}:{port}'.format(
+        http_code='{http_code}', host=host, port=port)
+    out = remote.check_call(cmd)['stdout'][0]
+    if out.isdigit() is True:
```
`is True` is not necessary here

alex7c4 replied:
I know, just from my point of view `is True` is more easy to read in the code

gdyuldin replied:
Let's compare `if out is digit is true` and `if out is digit`

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +235,32 @@ def consume_msg(remote, cfg_file_path):
+
+def rabbit_rpc_curl(remote, host, port=5000):
+    # curl to client
+    cmd = 'curl -w %{http_code} -s -o /dev/null http://{host}:{port}'.format(
```
`'curl -w %{{http_code}}'.format()` will return `'curl -w %{http_code}'`.
Is it what you want?

gdyuldin replied:
`requests.get(url).status_code` ?

alex7c4 replied:
> 'curl -w %{{http_code}}'.format() will return 'curl -w %{http_code}'.
> Is it what you want?

Thanks!

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +235,32 @@ def consume_msg(remote, cfg_file_path):
+    return remote.host
+
+
+def rabbit_rpc_curl(remote, host, port=5000):
```
`def get_http_code(remote, url)` ?

Commit:
fix comments

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +236,33 @@ def consume_msg(remote, cfg_file_path):
+
+def get_http_code(host_ip, port=5000):
+    # curl to client
+    url = 'http://' + host_ip + ":" + str(port)
```
`'http://{host_ip}:{port}'`

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -416,3 +443,84 @@ def test_load_messages_and_restart_prim_nonprim_ctrlr(restart_ctrlr, env):
+
+    # node -> client: Check GET before controller(s) restart
+    node = random.choice(all_nodes)
+    with node.ssh() as remote:
```
Let's remove remote

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +236,33 @@ def consume_msg(remote, cfg_file_path):
+        status_code = requests.get(url).status_code
+        return status_code
+    except Exception as e:
+        return e
```
What you do with returned exception later?

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -416,3 +443,84 @@ def test_load_messages_and_restart_prim_nonprim_ctrlr(restart_ctrlr, env):
+    node = random.choice(all_nodes)
+    with node.ssh() as remote:
+        logger.debug('GET: [{0}] -> [{1}]'.format(remote.host, rpc_client_ip))
+        assert get_http_code(rpc_client_ip) == exp_resp
```
you don't use node or remote here.

rpromyshlennikov replied:
He uses it here:
logger.debug('GET: [{0}] -> [{1}]'.format(remote.host, rpc_client_ip))
But we can get host (hostname) directly from node.

gdyuldin replied:
Actually we make get request from server, not from remote host. So this line in record will lie.

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +236,33 @@ def consume_msg(remote, cfg_file_path):
+    background = '<&- >/dev/null 2>&1 &'
+    cmd = 'oslo_msg_check_server --nodebug --config-file {0} {1}'.format(
+        cfg_file_path, background)
+    remote.execute(cmd)
```
AFAIK, this tests marked as undestructive, so you need to stop this processes after test.

alex7c4 replied:
Done.
I've added "fixt_kill_rpc_server_client" fixture

gdyuldin commented on mos_tests/rabbitmq_oslo/test_restarts.py:
```diff
@@ -236,6 +236,33 @@ def consume_msg(remote, cfg_file_path):
+    background = '<&- >/dev/null 2>&1 &'
+    cmd = 'oslo_msg_check_client --nodebug --config-file {0} {1}'.format(
+        cfg_file_path, background)
+    remote.execute(cmd)
```
Same, as above

alex7c4 replied:
Done.
I've added "fixt_kill_rpc_server_client" fixture

Commit:
Add QA-2293, QA-2294

QA-2293 - Start RabbitMQ RPC server and client and restart non-primary controller
QA-2294 - Start RabbitMQ RPC server and client and restart primary controller
